### PR TITLE
rgw/multisite: disable mdlog async notifications

### PR DESCRIPTION
**Pull request by smanjara**
With async notifications enabled, zones would excessively poll master zone to clone metadata log preventing the 20 second sleep interval between the RGWCloneMetaLogCoroutine() calls. This pr disables mdlog async notifications. They are not required for a successful metadata sync.

https://tracker.ceph.com/issues/61743

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by mattbenjamin — COMMENTED** [review]
so, this disables notifications by default, not altogether;  is that the consensus?

**Comment by smanjara** [comment]
> so, this disables notifications by default, not altogether; is that the consensus?

that's right, the code can still be exercised with rgw_md_notify_interval_msec > 0

**Comment by cbodley** [comment]
> With async notifications enabled, zones would excessively poll master zone to clone metadata log preventing the 20 second sleep interval between the RGWCloneMetaLogCoroutine() calls.

@smanjara isn't this the intent of async notifications though? the wakeups mean that there's something new to poll for

do we have evidence from the customer case that notifications are involved on the shards in question?

**Comment by smanjara** [comment]
> > With async notifications enabled, zones would excessively poll master zone to clone metadata log preventing the 20 second sleep interval between the RGWCloneMetaLogCoroutine() calls.
> 
> @smanjara isn't this the intent of async notifications though? the wakeups mean that there's something new to poll for

@cbodley that's right, but we enter a continous loop after, where the polling doesn't stop even when there's nothing new to poll. 

> 
> do we have evidence from the customer case that notifications are involved on the shards in question?

no, but with my testing I have observed that there is a single wakeup call when async notification kicks in after which the continous loop reproduces, which seems similar to the case.

**Comment by mkogan1** [comment]
TL;DR: 
a buckets delete workload reproduces the issue clearly (the create also reproduces it but to lesser extent)
and this PR fixes it.


_Repro Environment Details:_
- MS setup with 3x3 RGWs dedicated to MS sync, and 1x1 RGWs for client load
- the branch is `main` with this PR commit cherry-pick  (**has the sync fairness merged**)
`* main                                                     107e0a9877d [origin/main: ahead 1] rgw/multisite: disable mdlog async notifications to prevent excessive po`

- workload flow: 
1. create 100'000 buckets on the primary, wait for sync 'caught up'
2. delete the 100'000 buckets from the primary, wait for sync 'caught up'
3. monitor the **metadata** polling rate on primary MS sync RGWs logs before, during and after the loads


_Results:_

1. **without the fix:**
`-o rgw_md_notify_interval_msec=200`

// create 100K buckets, highlighted in purple is the polling rate during the bucket creating:
```
❯ time nice ~/go/bin/hsbench -a b2345678901234567890 -s b234567890123456789012345678901234567890 -u http://127.0.0.1:8000 -z 4K -d -1 -t $(nproc) -b 100000 -n 0 -m ip -bp hsbench-1- -op useast_stage1_ |& tee hsbench.log | stdbuf -oL -eL colrm 170 | ccze -Aonolookups
❯ rg --text --no-heading --no-filename '^202' ./out/radosgw.800{1..3}.log | pv -s $(du -cb ./out/radosgw.800{1..3}.log | awk '{print $1}' | tail -1) -cN 'rg' | sed 's/T/ /' | nice sort -k2,2 -S1G -o ./rgws-01.log ; cat ./rgws-01.log | rg 'beast' | rg 'type=metadata' | m-lnav
```
after the sync caught up the polling rate stabilizes on a higher rate:
image


// delete the 100K buckets, highlighted in purple is the polling rate during the bucket deletion:
```
❯ time nice ~/go/bin/hsbench -a b2345678901234567890 -s b234567890123456789012345678901234567890 -u http://127.0.0.1:8000 -z 4K -d -1 -t $(nproc) -b 100000 -n 0 -m cx -bp hsbench-1- -op useast_stage1_ |& tee hsbench.log | stdbuf -oL -eL colrm 170 | ccze -Aonolookups
❯ rg --text --no-heading --no-filename '^202' ./out/radosgw.800{1..3}.log | pv -s $(du -cb ./out/radosgw.800{1..3}.log | awk '{print $1}' | tail -1) -cN 'rg' | sed 's/T/ /' | nice sort -k2,2 -S1G -o ./rgws-01.log ; cat ./rgws-01.log | rg 'beast' | rg 'type=metadata' | m-lnav
```
after the sync caught up the polling rate stabilizes on about double the poll op/sec (2300 after .vs. 1200 before):
image


2. **WITH the fix:**
`-o rgw_md_notify_interval_msec=0`

// create 100K buckets, highlighted in purple is the polling rate during the bucket creating:
```
❯ time nice ~/go/bin/hsbench -a b2345678901234567890 -s b234567890123456789012345678901234567890 -u http://127.0.0.1:8000 -z 4K -d -1 -t $(nproc) -b 100000 -n 0 -m ip -bp hsbench-1- -op useast_stage1_ |& tee hsbench.log | stdbuf -oL -eL colrm 170 | ccze -Aonolookups
❯ rg --text --no-heading --no-filename '^202' ./out/radosgw.800{1..3}.log | pv -s $(du -cb ./out/radosgw.800{1..3}.log | awk '{print $1}' | tail -1) -cN 'rg' | sed 's/T/ /' | nice sort -k2,2 -S1G -o ./rgws-01.log ; cat ./rgws-01.log | rg 'beast' | rg 'type=metadata' | m-lnav
```
after the sync caught up the polling rate stabilizes on a same rate:
image


// delete the 100K buckets, highlighted in purple is the polling rate during the bucket deletion:
```
❯ time nice ~/go/bin/hsbench -a b2345678901234567890 -s b234567890123456789012345678901234567890 -u http://127.0.0.1:8000 -z 4K -d -1 -t $(nproc) -b 100000 -n 0 -m cx -bp hsbench-1- -op useast_stage1_ |& tee hsbench.log | stdbuf -oL -eL colrm 170 | ccze -Aonolookups
❯ rg --text --no-heading --no-filename '^202' ./out/radosgw.800{1..3}.log | pv -s $(du -cb ./out/radosgw.800{1..3}.log | awk '{print $1}' | tail -1) -cN 'rg' | sed 's/T/ /' | nice sort -k2,2 -S1G -o ./rgws-01.log ; cat ./rgws-01.log | rg 'beast' | rg 'type=metadata' | m-lnav
```
after the sync caught up the polling rate stabilizes on a same rate:
image

**Review by mkogan1 — APPROVED** [review]

**Comment by cbodley** [comment]
thanks @smanjara, it's good that we have a consistent reproducer now. but this pr seems more like a workaround than a fix. we discussed potential fixes to the looping logic over email. can you use this reproducer to test those changes?

**Comment by smanjara** [comment]
> thanks @smanjara, it's good that we have a consistent reproducer now. but this pr seems more like a workaround than a fix. we discussed potential fixes to the looping logic over email. can you use this reproducer to test those changes?



> TL;DR: a buckets delete workload reproduces the issue clearly (the create also reproduces it but to lesser extent) and this PR fixes it.
> 

thank you @mkogan1

**Comment by smanjara** [comment]
> @smanjara right!
> 
yeah I tested that along with the notifications disabled. I missed creating a pr for it.

> regarding this PR, it's good to support `rgw_md_notify_interval_msec=0` as a way to disable the notifications, but i don't think we should change the default value
> 
> at some point, i'd like to revive #46518 to remove the feature entirely

thanks! I wasn't aware of it. will test and see if it needs any more changes.

**Comment by mkogan1** [comment]
@smanjara as discussed checked `rgw_md_notify_interval_msec=200` and reverting of
https://github.com/ceph/ceph/commit/b0ce572b6f59c91b9c202ade076dbb90105f40a0
polling rate after buckets delete is very high and subsiding very slowly
image

**Comment by mkogan1** [comment]
@smanjara @cbodley @mattbenjamin 
another data point - restarting the secondary zones RGW's stops the polling in the primary RGWs logs:
image
